IPython/terminal/pt_inputhooks/wx.py

```diff
@@ -90,58 +107,115 @@ def inputhook_wx3(context):
-            del ea
-    except KeyboardInterrupt:
-        pass
+    app = wx.GetApp()
+    if app is not None:
+        assert wx.Thread_IsMain()
+
+        # The import of wx on Linux sets the handler for signal.SIGINT
+        # to 0.  This is a bug in wx or gtk.  We fix by just setting it
+        # back to the Python default.
+        if not callable(signal.getsignal(signal.SIGINT)):
+            signal.signal(signal.SIGINT, signal.default_int_handler)
+
+        evtloop = wx.EventLoop()
+        ea = wx.EventLoopActivator(evtloop)
+        t = clock()
+        while not context.input_is_ready():
+            while evtloop.Pending():
+                t = clock()
+                evtloop.Dispatch()
+            app.ProcessIdle()
+            # We need to sleep at this point to keep the idle CPU load
+            # low.  However, if sleep to long, GUI response is poor.  As
+            # a compromise, we watch how often GUI events are being processed
+            # and switch between a short and long sleep time.  Here are some
+            # stats useful in helping to tune this.
+            # time    CPU load
+            # 0.001   13%
+            # 0.005   3%
+            # 0.01    1.5%
+            # 0.05    0.5%
+            used_time = clock() - t
+            if used_time > 10.0:
+                # print 'Sleep for 1 s'  # dbg
+                time.sleep(1.0)
+            elif used_time > 0.1:
+                # Few GUI events coming in, so we can sleep longer
+                # print 'Sleep for 0.05 s'  # dbg
+                time.sleep(0.05)
+            else:
+                # Many GUI events coming in, so sleep only very little
+                time.sleep(0.001)
+        del ea
     return 0
 
-if sys.platform == 'darwin':
-    # On OSX, evtloop.Pending() always returns True, regardless of there being
-    # any events pending. As such we can't use implementations 1 or 3 of the
-    # inputhook as those depend on a pending/dispatch loop.
+
+@ignore_keyboardinterrupts
+def inputhook_wxphoenix(context):
+    """Run the wx event loop until the user provides more input.
+
+    This input hook is suitable for use with wxPython >= 4 (a.k.a. Phoenix).
+
+    It uses the same approach to that used in
+    ipykernel.eventloops.loop_wx. The wx.MainLoop is executed, and a wx.Timer
+    is used to periodically poll the context for input. As soon as input is
+    ready, the wx.MainLoop is stopped.
+    """
+
+    app = wx.GetApp()
+
+    if app is None:
+        return
+
+    if context.input_is_ready():
+        return
+
+    assert wx.IsMainThread()
+
+    # Wx uses milliseconds
+    poll_interval = 100
+
+    # We have to create a dummy wx.Frame, otherwise wx.App.MainLoop will know
+    # that it has nothing to do, and will return immediately.
+    frame = getattr(inputhook_wxphoenix, '_frame', None)
+    if frame is None:
+        inputhook_wxphoenix._frame = frame = wx.Frame(None)
+        frame.Show(False)
```

Comment:
You may be able to avoid the need for a frame if you use `app.SetExitOnFrameDelete(False)`

Reply:
I think Matplotlib sets that on the app so that if you do call `app.MainLoop()` to block the process and then close all of the plots you get the prompt back.

That said, if there are no open windows at all, and we are waiting for user input is it OK that this short-circuits and immediately returns?

Reply:
@RobinD42 yep, that seems to work fine, thanks for the tip!

@tacaswell It doesn't seem to be a problem; I'll do some more testing over the next few days (away from work, so have limited time right now).